### PR TITLE
Optimized FastExpressionCompiler output

### PR DESCRIPTION
**Pull request by Havunen**
Optimized FastExpressionCompiler emitted code. When expressions are large; push closure.items and constants into local variables to reduce memory consumption and easy reuse.

Previously LoadTest type index: 148 emitted 530 962 IL operations.
This change reduces emitted code to 177 942 IL operations. 
For reference same type compiled using Lambda Compiler emits 188 669 IL operations.

Compiled version of load test finishes now in 17seconds, before this change: 45seconds.

When there is less than 4 constants we use the old routine.
- load arg
- read field
- load constant
- read array index by constant

After this we optimize this routine by first creating locals for each constant so previous routine becomes.
- load local var

so the previous routine becomes 1 IL operation.

Here is code for reference.

Before:
```
IL_0000: ldarg.0    
IL_0001: ldfld      System.Object[] ConstantsAndNestedLambdas/FastExpressionCompiler.LightExpression.ExpressionCompiler+ArrayClosure
IL_0006: ldc.i4.0   
IL_0007: ldelem.ref 
IL_0008: ldarg.0    
IL_0009: ldfld      System.Object[] ConstantsAndNestedLambdas/FastExpressionCompiler.LightExpression.ExpressionCompiler+ArrayClosure
IL_000e: ldc.i4.1   
IL_000f: ldelem.ref 
IL_0010: newobj     Void .ctor()/Pdf.PdfDocumentContainerService
IL_0015: newobj     Void .ctor(Pdf.IPdfDocumentContainerService)/Pdf.PdfDocumentService
IL_001a: ldarg.0    
IL_001b: ldfld      System.Object[] ConstantsAndNestedLambdas/FastExpressionCompiler.LightExpression.ExpressionCompiler+ArrayClosure
IL_0020: ldc.i4.2   
IL_0021: ldelem.ref 
IL_0022: ldarg.0    
IL_0023: ldfld      System.Object[] ConstantsAndNestedLambdas/FastExpressionCompiler.LightExpression.ExpressionCompiler+ArrayClosure
IL_0028: ldc.i4.3   
IL_0029: ldelem.ref 
IL_002a: ldarg.0    
IL_002b: ldfld      System.Object[] ConstantsAndNestedLambdas/FastExpressionCompiler.LightExpression.ExpressionCompiler+ArrayClosure
IL_0030: ldc.i4.4   
IL_0031: ldelem.ref 
IL_0032: ldarg.0    
IL_0033: ldfld      System.Object[] ConstantsAndNestedLambdas/FastExpressionCompiler.LightExpression.ExpressionCompiler+ArrayClosure
IL_0038: ldc.i4.5   
IL_0039: ldelem.ref 
IL_003a: ldarg.0    
IL_003b: ldfld      System.Object[] ConstantsAndNestedLambdas/FastExpressionCompiler.LightExpression.ExpressionCompiler+ArrayClosure
IL_0040: ldc.i4.6   
IL_0041: ldelem.ref 
IL_0042: ldarg.0    
IL_0043: ldfld      System.Object[] ConstantsAndNestedLambdas/FastExpressionCompiler.LightExpression.ExpressionCompiler+ArrayClosure
IL_0048: ldc.i4.7   
IL_0049: ldelem.ref 
IL_004a: ldarg.0    
IL_004b: ldfld      System.Object[] ConstantsAndNestedLambdas/FastExpressionCompiler.LightExpression.ExpressionCompiler+ArrayClosure
IL_0050: ldc.i4.8   
IL_0051: ldelem.ref 
IL_0052: ldarg.0    
IL_0053: ldfld      System.Object[] ConstantsAndNestedLambdas/FastExpressionCompiler.LightExpression.ExpressionCompiler+ArrayClosure
IL_0058: ldc.i4.s   9
IL_005a: ldelem.ref 
IL_005b: ldarg.0    
IL_005c: ldfld      System.Object[] ConstantsAndNestedLambdas/FastExpressionCompiler.LightExpression.ExpressionCompiler+ArrayClosure
IL_0061: ldc.i4.s   10
IL_0063: ldelem.ref 
IL_0064: ldarg.0    
IL_0065: ldfld      System.Object[] ConstantsAndNestedLambdas/FastExpressionCompiler.LightExpression.ExpressionCompiler+ArrayClosure
IL_006a: ldc.i4.s   11
IL_006c: ldelem.ref 
IL_006d: newobj     Void .ctor(Framework.IContextService`1[Shared.IPsaContext], Data.IBusinessUnitRepository, Shared.IOrganizationCompanyRepository)/Logic.BusinessUnitValidator
```
After:
```
IL_1429: ldloc.1    
IL_142a: ldloc.2    
IL_142b: ldloc.3    
IL_142c: ldloc.s    V_4
IL_142e: ldloc.s    V_5
IL_1430: ldloc.s    V_6
IL_1432: ldloc.s    V_7
IL_1434: ldloc.s    V_8
IL_1436: ldloc.s    V_9
IL_1438: ldloc.s    V_10
IL_143a: ldloc.s    V_11
IL_143c: newobj     Void .ctor(Framework.IContextService`1[Shared.IPsaContext], Data.IBusinessUnitRepository, Shared.IOrganizationCompanyRepository)/Logic.BusinessUnitValidator
```

### CODE REVIEW

**Comment by Havunen** [comment]
This optimization will not however totally fix https://github.com/dadhi/DryIoc/issues/139 because even larger expressions than LoadTest will still cause stackoverflow exception.

File: src/DryIoc/FastExpressionCompiler.cs [hunk 1]
```diff
@@ -2169,19 +2175,53 @@ private static bool TryEmitConstant(ConstantExpression expr, Type exprType, obje
...
                         return false;
 
-                    // Load constant from Closure - closure object is always a first argument
-                    il.Emit(OpCodes.Ldarg_0);
-                    il.Emit(OpCodes.Ldfld, ArrayClosure.ConstantsAndNestedLambdasField);
-                    EmitLoadConstantInt(il, constIndex);
-                    il.Emit(OpCodes.Ldelem_Ref);
-                    il.Emit(exprType.IsValueType() ? OpCodes.Unbox_Any : OpCodes.Castclass, exprType);
+                    // If expression is small and there is no variables set then just read them by args and field
+                    if (constantCount <= 3 && closure.ConstantVarsRef == null)
```
Comment (Havunen) [inline src/DryIoc/FastExpressionCompiler.cs]: 3 is magic number, until variables are taken into use. I don't know if it should be more or less.

**Comment by Havunen** [comment]
Its a tool called IL Visualizer and I use its Visual Studio extension. It did not work out of the box so I forked the open source project and built it for .Net 48 and visual studio 2019
My fork is here: https://github.com/Havunen/il-visualizer

**Comment by dadhi** [comment]
> Its a tool called IL Visualizer
Ah, then it is the same as I use. But I need to change target to Net 4.5 each time to make it work.

**Comment by dadhi** [comment]
@Havunen,

So you storing constants in locals once, and then load on each usage.
I like it. Also liked the optimization with emitting `_S` opcodes.

- To proceed we need to move the changes to FEC repo
- Ensure that all tests are passed
- Benchmark that adding LocalBuilders to closure and the new logic 
did not affect (significally) compilation, invocation and memory. We can use existing benchmarks listed in readme and contained in `..Benchmarks` project.
- Cleanup or revert formatting changes cause it harder to spot actual changes.
- After merging to FEC we can merge to DryIoc master

**Comment by dadhi** [comment]
Meanwhile I had another idea which does not involve the locals:
to add `ArrayClosure.Const1..N` properties returning `=> Items[1..N]`.
So it will be closure arg log then property call opcodes.

I am not sure how adding variables increase the stack size,
so using the properties maybe a good tradeoff. The logic will be simpler as well.

I have very limited time this week, so not sure whether I can try it.

**Comment by Havunen** [comment]
Ok. Lets continue there.

**Pull request closed without merging**